src/OVAL/oval_session.c

```diff
@@ -0,0 +1,463 @@
+#include <config.h>
+#endif
+
+#include <libgen.h>
+#include <unistd.h>
+#include <string.h>
+
+#include "common/alloc.h"
+#include "common/util.h"
+#include "common/_error.h"
+#include "common/oscapxml.h"
+#include "source/xslt_priv.h"
+#include "public/oval_agent_api.h"
+#include "public/oval_session.h"
+#include "../DS/public/ds_sds_session.h"
+#include "oscap_source.h"
+
+static const char *oscap_productname = "cpe:/a:open-scap:oscap";
+static const char *oval_results_report = "oval-results-report.xsl";
+
+struct oval_session {
+	/* Main source assigned with the main file (SDS or OVAL) */
+	struct oscap_source *source;
+	struct oval_definition_model *def_model;
+	struct oval_variable_model *var_model;
+	struct oval_results_model *res_model;
+
+	oval_agent_session_t *sess;
+	struct ds_sds_session *sds_session;
+
+	struct {
+		struct oscap_source *definitions;
+		struct oscap_source *variables;
+		struct oscap_source *directives;
+	} oval;
+
+	const char *datastream_id;
+	/* particular OVAL component if there are two OVALs in one datastream */
+	const char *component_id;
+
+	struct {
+		const char *results;
+		const char *report;
+	} export;
+
+	struct {
+		/* it's called when there is something invalid in input/output files */
+		xml_reporter xml_fn;
+	} reporter;
+
+	bool validation;
+	bool full_validation;
+};
+
+struct oval_session *oval_session_new(const char *filename)
+{
+	oscap_document_type_t scap_type;
+	struct oval_session *session;
+
+	session = (struct oval_session *) oscap_calloc(1, sizeof(struct oval_session));
+
+	session->source = oscap_source_new_from_file(filename);
+	if ((scap_type = oscap_source_get_scap_type(session->source)) == 0) {
+		oval_session_free(session);
+		return NULL;
+	}
+
+	if (scap_type != OSCAP_DOCUMENT_OVAL_DEFINITIONS && scap_type != OSCAP_DOCUMENT_SDS) {
+		oscap_seterr(OSCAP_EFAMILY_OSCAP, "Session input file was determined but it"
+				" isn't an OVAL file nor a source datastream file.");
+		oval_session_free(session);
+		return NULL;
+	}
+
+	return session;
+}
+
+void oval_session_set_variables(struct oval_session *session, const char *filename)
+{
+	__attribute__nonnull__(session);
+
+	oscap_source_free(session->oval.variables);
+
+	if (filename != NULL)
+		session->oval.variables = oscap_source_new_from_file(filename);
+	else
+		session->oval.variables = NULL;	/* reset */
+}
+
+void oval_session_set_directives(struct oval_session *session, const char *filename)
+{
+	__attribute__nonnull__(session);
+
+	oscap_source_free(session->oval.directives);
+
+	if (filename != NULL)
+		session->oval.directives = oscap_source_new_from_file(filename);
+	else
+		session->oval.directives = NULL;
+}
+
+void oval_session_set_validation(struct oval_session *session, bool validate, bool full_validation)
+{
+	__attribute__nonnull__(session);
+
+	session->validation = validate;
+	session->full_validation = full_validation;
+}
+
+void oval_session_set_datastream_id(struct oval_session *session, const char *id)
```

Comment:
My opinion is not use macros to generate any functions, because with macros I always got confused while debugging. I think it is not a problem that there are many setters with similar code.